Add support for statsd export

douglas-reid opened the pull request:
The goal of this PR is to enable collection of proxy stats. This PR focuses on generating envoy configs to direct statsd output.

This PR updates the proxy agent and the manager to support a new field in the mesh config: `statsd_udp_address`. The initial default option is to generate this address from the IP address of Mixer and the configured statsd-udp port (defaults to 9125).

Tested on new deployments. A PR to istio/istio that updates the deployment configs for mixer, pilot, and prometheus will follow.

This has a current weakness: if mixer service is deleted and restarted, export will stop working, as the `statsd_udp_address` is not regenerated in response.

kyessenov commented:
There is a linter warning (run `bin/check.sh`)

kyessenov commented on model/validation.go:
```diff
@@ -806,6 +806,28 @@ func ValidateProxyAddress(hostAddr string) error {
 }
 
+// ValidateIPv4AddressAndPort checks that a network address is well-formed.
+// This works just like ValidateProxyAddress, but without support for FQDNs.
```
why is FQDN not supported?

douglas-reid replied:
This was needed because the statsd udp address (as currently processed by Envoy) has to be <IP>:<Port>. So, the thinking was to have some validation of something that can only be an IP Addr.

kyessenov replied:
I see

kyessenov commented on cmd/cmd.go:
```diff
@@ -59,6 +60,24 @@ func GetMeshConfig(kube kubernetes.Interface, namespace, name string) (*proxycon
+		if err := model.ValidateIPv4Address(mixerAddr); err == nil {
+			mesh.StatsdUdpAddress = mesh.MixerAddress
+		} else {
+			if mixerSvc, err := kube.CoreV1().Services(namespace).Get(mixerAddr, v1.GetOptions{}); err == nil {
```
cmd/ should not be depending on k8s. I'd prefer we did not automatically set values in the config since it obscures problems and makes debugging difficult.

kyessenov replied:
(regarding the fetch above, it is going away once mesh config becomes a proper config resource)

douglas-reid replied:
Without this, it makes configuring the export much harder. I'm open to good ideas, however.

Envoy wants an IP, not a service name. So, to get this to work, something needs to map from say mixer_server to ip. It seems easier to have cmd do this than it is to have an operator deploy mixer first, get the IP, then update the mesh file, then deploy pilot.

But, yes, if there is a better way, let's shift to that.

kyessenov replied:
Can you do DNS resolution yourself here? Please don't fallback onto mixer address though, we should just change the value of statsd address to IP:PORT until Envoy gets DNS support for statsd setting. You can use go's `LookupHost` to parse out and replace host.

istio-testing commented:
Jenkins job pilot/presubmit passed

istio-testing commented:
Jenkins job pilot/e2e-smoketest passed

istio-testing commented:
Jenkins job pilot/presubmit passed

kyessenov commented on proxy/envoy/config.go:
```diff
@@ -111,6 +115,21 @@ func buildConfig(listeners Listeners, clusters Clusters, mesh *proxyconfig.Proxy
 		},
 	}
 
+	if mesh.StatsdUdpAddress != "" {
```
config gen is supposed to be a pure function since its output is cached. Can you please move this chunk to `proxy/envoy/watcher.go` close to where agent is instantiated? You'd also need to resolve DNS in `proxy/envoy/ingress.go` and `proxy/envoy/egress.go`. 

If you want to reload Envoy on DNS change, you'd need to add a polling routine and trigger. I suggest not do that and just wait till Envoy starts using UDP clusters.

rshriram replied:
@douglas-reid silly question: why the choice of udp? why not use TCP?

douglas-reid replied:
@rshriram the idea is to take advantage of the `statsd_udp_ip_address` field recently added to envoy (in conjunction with the official https://github.com/prometheus/statsd_exporter). This is all to quickly get some (shared) unified visibility into the envoy metrics.

douglas-reid commented:
@kyessenov I've updated to do resolve before agent instantiation and added to ingress/egress per your comments. I believe this matches what you were asking for. PTAL.

kyessenov commented on proxy/envoy/egress.go:
```diff
@@ -41,6 +41,14 @@ func NewEgressWatcher(mesh *proxyconfig.ProxyMeshConfig) (Watcher, error) {
+		if addr, err := resolveStatsdAddr(mesh.StatsdUdpAddress); err == nil {
+			mesh.StatsdUdpAddress = addr
+		} else {
+			glog.V(2).Infof("Error resolving statsd address; clearing to prevent bad config: %v", err)
```
Shouldn't this be a warning?

douglas-reid replied:
done.

istio-testing commented:
Jenkins job pilot/presubmit passed

istio-testing commented:
Jenkins job pilot/presubmit passed

istio-testing commented:
Jenkins job pilot/e2e-smoketest passed

kyessenov commented:
Thanks!

istio-testing commented:
Jenkins job pilot/presubmit passed

istio-testing commented:
Jenkins job pilot/e2e-smoketest passed

douglas-reid commented:
Thanks for the good review, and the merge!
